File: inter_detect/src/inter_detect_node.py
#! /usr/bin/python
# it requires "chmod +x mypythonscript.py" to be called by ROS
##############################################################
# Intersection line detector v1
# Autor: Cesar Bravo
# 29/03/2017
##############################################################

import rospy
from sensor_msgs.msg import CompressedImage
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError
from std_msgs.msg import Int16
import cv2
import numpy as np
import time
import math
from pendCalc import pend
from numpy import *

#Constant
LINE_WIDTH = 8
TOL = 4

bridge = CvBridge()


def image_callback(ros_data):
	
	
    try:
        # Convert your ROS Image message to OpenCV2

	# Uncomment the following line to work with raw images
        #cv2_img = bridge.imgmsg_to_cv2(ros_data, "bgr8")

	# Uncoment the following two lines to work with compressed images
	np_arr = np.fromstring(ros_data.data, np.uint8)
	cv2_img = cv2.imdecode(np_arr, cv2.CV_LOAD_IMAGE_GRAYSCALE)

    except CvBridgeError, e:
        print(e)
    else:

	global dist
	global ang
	# Processing function begins
	e1 = cv2.getTickCount()

        # 1. Convert data received to numpy array
        image = np.asarray(cv2_img) # 480x640x3
	scld = cv2.resize(image,None,fx=0.6,fy=0.6,interpolation=cv2.INTER_AREA)

	# 2. Filtering and Binarization ----------------------------------------
	thd = cv2.inRange(scld, 140, 190)
	hght, wdth = thd.shape[:2] #216x192
	#print(hght,wdth)

	# 3. Intersection line detection ---------------------------------------
	i = hght-2
	j = int(wdth*(0.35))

	point = hght-1
	ini = []

	while j < wdth*(0.7):	

		cont = 0
		pc = []
		#Lado Izquierdo
		while hght*(0.5) < i < hght:
			if thd[i,j] != 0:
				cont += 1
				pc.append(i)

			else:
				if LINE_WIDTH-TOL < cont < LINE_WIDTH+TOL:
					if abs(point - pc[0])<7:
						ini.append((pc[0],j))
						cv2.circle(scld,(j,pc[0]),1,250,-1)
						point = pc[0]+TOL
						i = 0
					else:
						point = pc[0] + TOL
						i = 0
				cont = 0
				pc = []		

			i -= 1

		i = hght - 5
		j += 2		
			
	# Distance and angle calculation and publication -----------------------
	p = None
	if len(ini)>10:
		p = pend(ini)

	else: 
		dist = 200

	if p != None:

		x1 = linspace(wdth*0.25,wdth*0.7,10)
		y1 = poly1d(p)
		for i in x1:

			cv2.circle(scld,(int(i),int(y1(i))),1,0,-1)
		
		thtCalc = int(57*arctan(p[0]))
		ang = thtCalc
		dist = int(hght - int(mean(ini, axis=0)[0]))

	if ang>-15 and ang<15:

		pubDist.publish(dist)
		pubAng.publish(ang)
	

	
	# Compress image to pub ------------------------------------------------
        cropImage = CompressedImage()
        cropImage.header.stamp = rospy.Time.now()
        cropImage.format = "jpeg"
        cropImage.data = np.array(cv2.imencode('.jpg',scld)[1]).tostring()
        pub.publish(cropImage)

	# Print stats ----------------------------------------------------------
	e2 = cv2.getTickCount()	
    	t = (e2 - e1)/cv2.getTickFrequency()	
	
	print('angle: '+str(ang)+'| distance: '+str(dist)+'| frame time:'+str(t))
	print('----------------------------------------------------block end')
        

def main():

    global pub
    global pubDist
    global pubAng
    global ang
    global dist
   

    ang = 180
    dist = 100

    rospy.init_node('inter_line')
    image_topic = "/img_prepros/compressed"
    rospy.Subscriber(image_topic,CompressedImage,image_callback,queue_size=1)
    pub = rospy.Publisher('/inter_line/compressed',CompressedImage, queue_size=1)
    pubDist = rospy.Publisher('/inter_line/dist', Int16, queue_size=1)
    pubAng = rospy.Publisher('/inter_line/ang', Int16, queue_size=1)
    rospy.spin()

if __name__ == '__main__':
    main()

File: inter_detect/src/inter_detect_node.cpp
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/image_encodings.h>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <ros/console.h>
#include <math.h>
#include <numeric>
#include "std_msgs/Int16.h"

int LINE_WIDTH=8;
int TOL=4;

int dist;
int ang;
std_msgs::Int16 p_ang,p_dist;

class ImageConverter
{
  	ros::NodeHandle nh_;
  	image_transport::ImageTransport it_;
	image_transport::Subscriber image_sub_;
  	image_transport::Publisher image_pub_;
	ros::Publisher PubAng;
	ros::Publisher PubDist;

public:
  	ImageConverter()
    	: it_(nh_)
  {
    	image_sub_ = it_.subscribe("/img_prepros", 1,
      	&ImageConverter::imageCb, this,image_transport::TransportHints("compressed"));
    	image_pub_ = it_.advertise("/inter_line", 1);

	PubAng= nh_.advertise<std_msgs::Int16>("inter_line/ang",1);
	PubDist= nh_.advertise<std_msgs::Int16>("inter_line/dist",1);
  }

  void imageCb(const sensor_msgs::ImageConstPtr& msg)
  {
    	cv_bridge::CvImagePtr cv_ptr;
    	sensor_msgs::ImagePtr sal_ptr;
    try
    {
      	cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::MONO8);
    }
    catch (cv_bridge::Exception& e)
    {
      	ROS_ERROR("cv_bridge exception: %s", e.what());
      	return;
    }

	int e1=cv::getTickCount();
//La imagen contenida en cv_ptr se guarda en una matriz 
   	cv::Mat image=cv_ptr->image;

//Reescala
   	cv:: Mat scld;
   	cv::resize(image, scld, cv::Size(), 0.6, 0.6,CV_INTER_LINEAR);

//Binarización
    cv::Mat bin;
    cv::inRange(scld,140,190,bin );

	int wdth, hght,i,j,point,cont;
	std::vector<cv::Point>  ini;
	std::vector<int> pc;
	wdth=bin.cols;
	hght=bin.rows;

	i=hght-2;
	j=wdth*0.35;
	point=hght-1;
	while (j<(wdth*0.8))
	{

		cont=0;
		pc.erase(pc.begin(),pc.end());
		while ((hght*0.5<i)&&(i<hght))
		{
			if (bin.at<uchar>(i,j)!=0)
			{
				cont++;
				pc.push_back(i);
			}
			else
			{
				if ((LINE_WIDTH-TOL<cont)&&(cont<LINE_WIDTH+TOL))
				{
					if (abs(point-pc[0])<7)
					{
						ini.push_back(cv::Point2f(j,pc[0]));
						cv::circle(scld,cv::Point(j,pc[0]),1,250,-1);
                                                
		
					}
					point = pc[0]+TOL;
					i = 0;
				}
				else
				{
					cont=0;
					pc.erase(pc.begin(),pc.end());
				}
				
			}
			i--;
		}
		i = point;
		j += 2;
	}
int x=ini.size();
if (ini.size()>13){
	cv::Vec4f p;
	cv::fitLine(ini, p,CV_DIST_L1,0,0.01,0.01);
	cv::circle(scld,cv::Point(p[2],p[3]),1,0,-1);
	cv::line(scld,cv::Point(p[2],p[3]),cv::Point(p[2]+p[0]*50,p[3]+p[1]*50),250);
	int ang=(57*atan(p[1]/p[0]));
	int dist=(hght-p[3]);	
	p_ang.data=ang;
	p_dist.data=dist;
	ROS_INFO("dist: %i	|| ang: %i",dist,ang);

}
else {
        p_dist.data=200;
        PubDist.publish(p_dist);
     }
if ((ang>-15)&&(ang<15))
{
	PubAng.publish(p_ang);
	PubDist.publish(p_dist);
}


//ROS_INFO("dist: %i       || ang: %i",dist,ang);
cv_bridge::CvImage out_msg;
out_msg.header   = msg->header; // Same timestamp and tf frame as input image
out_msg.encoding = sensor_msgs::image_encodings::MONO8; // Or whatever
out_msg.image    = scld; // Your cv::Mat
sal_ptr=out_msg.toImageMsg();
image_pub_.publish(sal_ptr);
int e2=cv::getTickCount();
float t=(e2-e1)/cv::getTickFrequency();
ROS_INFO("frame time: %f-------------------------------block end", t);
  }
};
int main(int argc, char** argv)
{
  ros::init(argc, argv, "inter_detect");
  ImageConverter ic;
  ros::spin();
  return 0;
}
